python_modules/dagster/dagster/_core/execution/context/invocation.py

```diff
@@ -226,53 +243,73 @@ def partition_key(self) -> str:
     def asset_partition_key_for_output(self, output_name: str = "result") -> str:
         return self.partition_key
 
-    def has_tag(self, key: str) -> bool:
-        raise DagsterInvalidPropertyError(_property_msg("has_tag", "method"))
-
-    def get_tag(self, key: str) -> str:
-        raise DagsterInvalidPropertyError(_property_msg("get_tag", "method"))
-
-    def get_step_execution_context(self) -> StepExecutionContext:
-        raise DagsterInvalidPropertyError(_property_msg("get_step_execution_context", "methods"))
-
     def bind(
         self,
-        op_def_or_invocation: Union[OpDefinition, PendingNodeInvocation[OpDefinition]],
+        op_def: OpDefinition,
+        pending_invocation: Optional[PendingNodeInvocation[OpDefinition]],
+        assets_def: Optional[AssetsDefinition],
+        config_from_args: Optional[Mapping[str, Any]],
+        resources_from_args: Optional[Mapping[str, Any]],
     ) -> "BoundOpExecutionContext":
-        op_def = (
-            op_def_or_invocation
-            if isinstance(op_def_or_invocation, OpDefinition)
-            else op_def_or_invocation.node_def
-        )
+        from dagster._core.definitions.resource_invocation import resolve_bound_config
 
-        _validate_resource_requirements(self._resource_defs, op_def)
+        if resources_from_args:
+            if self._resource_defs:
+                raise DagsterInvalidInvocationError(
+                    "Cannot provide resources in both context and kwargs"
+                )
+            resource_defs = wrap_resources_for_execution(resources_from_args)
+            # assumption: no context manager shenanigans can happen with resources passed by args
+            resources = build_resources(resource_defs, self.instance).__enter__()
+        elif assets_def and assets_def.resource_defs:
+            for key in sorted(list(assets_def.resource_defs.keys())):
+                if key in self._resource_defs:
+                    raise DagsterInvalidInvocationError(
+                        f"Error when invoking {str(assets_def)} resource '{key}' "
+                        "provided on both the definition and invocation context. Please "
+                        "provide on only one or the other."
+                    )
+            resource_defs = wrap_resources_for_execution(
+                {**self._resource_defs, **assets_def.resource_defs}
+            )
+            # how do we handle context manager scoping for these re-entered resources?
+            #  * error if there are any from either set?
+            #  * do some exit / re-enter stuff?
```

Comment:
the context chaining that we currently has issues leaving some state behind as new context were created. This impl makes us confront it more directly. Probably need to cook up some gnarly tests to help figure out whats the best path

Reply:
Having trouble visualizing this rn, can you walk through the problematic state?

Reply:
In the current implementation we make new instances of the resources (and `__enter__` them)  whenever the `replace_` methods are called. If you were trying to run a "test" with a real resource that spins up a cluster that would happen like 2 to 3 times depending on if your `@asset` had resources and config due to that repeated initialization.

In this PR that happens just once at `bind` time, so we still have to potential needlessly double init an expensive resource. 

I thought the exit case was worse, but because the icky `__exit__` via `__del__` thing, I do think we tear down all the extra copies correctly.

Reply:
[3]